### PR TITLE
Swap lints

### LINKED ISSUE #682: lint assignment sequences in the form `a = b; b = a`
nicked from PVS-Studio – this can reuse most of the equality infrastructure in utils.rs

### LINKED ISSUE #714: Lint swapping/replacement using assignments.
This shoud be done using `std::mem::{swap, replace}` instead.

```rust
let mut a = 3;
let mut b = 4;
let tmp = a;
a = b;
b = tmp;
```

should generate a warning.

### PR DESCRIPTION
**Pull request by mcarton**
/me thinks @llogiq will say those are too easy.
/me thinks 2 down, 136 to go.

Fix  #714. Fix #682.

### CODE REVIEW

File: src/swap.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust,ignore
+/// a = b;
+/// b = a;
+/// ```
+declare_lint! {
+    pub SUSPICIOUS_SWAP,
```
Comment (llogiq) [inline src/swap.rs]: I think we should call this `SWAP_FAIL` or `ALMOST_SWAPPED`. The way it is it's not a swap.

**Comment by llogiq** [comment]
I'm just jealous that you have the time in front of your PC to write those lints. :smile:

In this case, to make it harder, the lints could ignore intermediate statements involving neither a nor b (nor t for manual_swap). But I'm fine with merging it as is, too.

File: src/swap.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        check_suspicious_swap(cx, block);
+    }
+}
+
+/// Implementation of the `MANUAL_SWAP` lint.
+fn check_manual_swap(cx: &LateContext, block: &Block) {
+    for w in block.stmts.windows(3) {
+        if_let_chain!{[
+            // let t = foo();
+            let StmtDecl(ref tmp, _) = w[0].node,
```
Comment (Manishearth) [inline src/swap.rs]: it might not be the first node.

**Comment by llogiq** [comment]
Thanks. 137 to go. :laughing:

**Pull request merged**